### PR TITLE
A simple RL model solving the CartPole problem from OpenAI gym.

### PR DESCRIPTION
**Pull request by mhong**
On a Linux GPU env, it took ~25 mini batches (batch size 16) to solve (i.e.,
trained to converged) with Adam optimizer. Hyper params such as learning rate have not been well tuned.

Some print statements (including commented out ones) are intentionally left
around for debugging purposes.

### CODE REVIEW

File: Gym/CartPole.swift [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Python
+import TensorFlow
+
+let np = Python.import("numpy")
+let gym = Python.import("gym")
+
+let HIDDEN_SIZE = 128
```
Comment (rxwei) [inline Gym/CartPole.swift]: Please use camel case for all variables.

File: Gym/CartPole.swift [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+let BATCH_SIZE = 16
+let PERCENTILE = 70
+
+func scalarFromPython<Scalar: PythonConvertible>(value: PythonObject) -> Scalar {
```
Comment (rxwei) [inline Gym/CartPole.swift]: This is not an idiomatic use. The caller should call the scalar type's initializer on the python object directly, and use force unwrapping.

File: Gym/CartPole.swift [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  let swiftValue = Scalar.init(value)
+  assert(swiftValue != nil)
+  return swiftValue!
+}
+
+func tensorFromPython<Scalar: NumpyScalarCompatible>(numpy: PythonObject) -> Tensor<Scalar> {
```
Comment (rxwei) [inline Gym/CartPole.swift]: Similarly, the caller should use `Tensor<Scalar>(numpy:)` directly.

File: Gym/CartPole.swift [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+func loss(model: Net, input: Tensor<Float>, target: Tensor<Float>)
+  -> Tensor<Float> {
+  let context = Context(learningPhase: .training)
+  let logits = model.applied(to: input, in: context)
+  return softmaxCrossEntropy(logits: logits, labels: target)
+}
+
+func train() {
+  let env = gym.make("CartPole-v0")
+  let obsSize: Int32 = scalarFromPython(value: env.observation_space.shape[0])
```
Comment (rxwei) [inline Gym/CartPole.swift]: Rename to `observationSize`.

File: Gym/CartPole.swift [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+  let tensor = Tensor<Scalar>(numpy: numpy)
+  assert(tensor != nil)
+  return tensor!
+}
+
+// A simple two layer dense net.
+public struct Net: Layer {
+  var l1, l2: Dense<Float>
+
+  public init(obsSize: Int, hiddenSize: Int, actionCount: Int) {
```
Comment (rxwei) [inline Gym/CartPole.swift]: Avoid short names. Use "observationSize".

File: Gym/CartPole.swift [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+func tensorFromPython<Scalar: NumpyScalarCompatible>(numpy: PythonObject) -> Tensor<Scalar> {
+  let tensor = Tensor<Scalar>(numpy: numpy)
+  assert(tensor != nil)
+  return tensor!
+}
+
+// A simple two layer dense net.
+public struct Net: Layer {
```
Comment (rxwei) [inline Gym/CartPole.swift]: This is a single file script, so we should drop all `public`s from this file.

File: Gym/CartPole.swift [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+// We pass in `net` by reference to avoid making a copy, even though this function is
+// running inferencing over net, and not mutating it.
+func nextBatch(env: PythonObject, net: inout Net, batchSize: Int, actionCount: Int32
+) -> [Episode] {
+  var obsNumpy = env.reset()
+
+  let context = Context(learningPhase: .inference)
+  var episodes: [Episode] = []
+  
+  // We build up a batch of observations and actions
```
Comment (rxwei) [inline Gym/CartPole.swift]: End comments with a period.

File: Gym/CartPole.swift [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+    // print("target tensor has shape \(target.shapeTensor)")
+
+    totalReward += episode.reward
+    retainedEpisodeCount += 1
+  }
+  
+  return (input, target, retainedEpisodeCount, totalReward / Float(retainedEpisodeCount))
+}
+
+// We pass in `net` by reference to avoid making a copy, even though this function is
```
Comment (rxwei) [inline Gym/CartPole.swift]: This is unnecessary and you should pass the net value directly instead. It's a constant, and there is completely no copy.

File: Gym/CartPole.swift [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+// running inferencing over net, and not mutating it.
+func nextBatch(env: PythonObject, net: inout Net, batchSize: Int, actionCount: Int32
```
Comment (rxwei) [inline Gym/CartPole.swift]: Use idiomatic argument labeling and line wrapping.

```swift
func nextBatch(
    inEnvironment env: PythonObject,
    using net: Net,
    batchSize: Int,
    actionCount: Int32
) -> [Episode] {
    ...
}
```

File: Gym/CartPole.swift [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+      } else {
+        obsNumpy = nextObs
+      }
+    }
+  }
+
+  return episodes
+}
+
+func loss(model: Net, input: Tensor<Float>, target: Tensor<Float>)
```
Comment (rxwei) [inline Gym/CartPole.swift]: This function should be inlined into `gradient(at:in:)` trailing closure. These function wrappers are trivial and indirections are unnecessary.

File: Gym/CartPole.swift [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    self.l1 = Dense<Float>(
+      inputSize: obsSize, outputSize: hiddenSize, activation: { $0 })
+
+    self.l2 = Dense<Float>(
+      inputSize: hiddenSize, outputSize: actionCount, activation: { $0 })
+  }
+
+  @differentiable(wrt: (self, input))
+  public func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
+    return l2.applied(to: relu(l1.applied(to: input, in: context)), in: context)
```
Comment (rxwei) [inline Gym/CartPole.swift]: Make `l1`'s activation be `relu` instead of calling `relu` here.

File: Gym/CartPole.swift [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+  struct Step {
+    let obs: Tensor<Float>
+    let action: Int32
+  }
+
+  let steps: [Step]
+  let reward: Float
+}
+
+func filterBatch(episodes: [Episode], actionCount: Int32
```
Comment (rxwei) [inline Gym/CartPole.swift]: Wrap lines correctly like `nextBatch`.
Reference: https://google.github.io/swift/

Comment (rxwei) [inline Gym/CartPole.swift]: This function does not have side effects and should be named like a noun. Something like `filteringBatch` is good.

File: Gym/CartPole.swift [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+  }
+}
+
+// An episode is a list of steps, where each step records the observation from
+// env and the action taken. They will serve respectively as the input and
+// target (label) of the neural net training.
+// 
+struct Episode {
+  struct Step {
+    let obs: Tensor<Float>
```
Comment (rxwei) [inline Gym/CartPole.swift]: Avoid using short names. Use `observations`.

File: Gym/CartPole.swift [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+      inputSize: hiddenSize, outputSize: actionCount, activation: { $0 })
+  }
+
+  @differentiable(wrt: (self, input))
+  public func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
+    return l2.applied(to: relu(l1.applied(to: input, in: context)), in: context)
+  }
+}
+
+// An episode is a list of steps, where each step records the observation from
```
Comment (rxwei) [inline Gym/CartPole.swift]: Please start doc comments with 3 slashes.

File: Gym/CartPole.swift [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+func loss(model: Net, input: Tensor<Float>, target: Tensor<Float>)
+  -> Tensor<Float> {
+  let context = Context(learningPhase: .training)
+  let logits = model.applied(to: input, in: context)
+  return softmaxCrossEntropy(logits: logits, labels: target)
+}
+
+func train() {
```
Comment (rxwei) [inline Gym/CartPole.swift]: Why wrap things in a `train` function and call it immediately after? This is not very idiomatic.

File: Gym/CartPole.swift (hunk 1 above)
Comment (mhong) [reply Gym/CartPole.swift]: Done

File: Gym/CartPole.swift (hunk 2 above)
Comment (mhong) [reply Gym/CartPole.swift]: The concern with force unwrapping is that it can be hard to debug. assert failure will point to the source location. Any suggestions?

File: Gym/CartPole.swift (hunk 6 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 5 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 11 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 14 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 13 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 12 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 8 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 9 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 7 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 10 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 15 above)
Comment (mhong) [reply Gym/CartPole.swift]: moved the code to top level.

File: Gym/CartPole.swift (hunk 4 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 2 above)
Comment (mhong) [reply Gym/CartPole.swift]: Discussed and converged on creating an `Optional.unwrapped()` extension. Thanks for the ideas and code!

File: Gym/CartPole.swift [hunk 16]
```diff
@@ -24,20 +24,15 @@ let batchSize = 16
...
-  return swiftValue!
-}
-
-func tensorFromPython<Scalar: NumpyScalarCompatible>(numpy: PythonObject) -> Tensor<Scalar> {
-  let tensor = Tensor<Scalar>(numpy: numpy)
-  // Before force-unwrapping, we use `assert()` to make unwrapping nil more
-  // debuggable (assert points to the source location.)
-  assert(tensor != nil)
-  return tensor!
+/// Force unwrapping with ! does not provide source location when unwrapping
```
Comment (rxwei) [inline Gym/CartPole.swift]: Use 2 slashes for non-doc-comments.

File: Gym/CartPole.swift [hunk 17]
```diff
@@ -134,34 +129,31 @@ func nextBatch(env: PythonObject,
...
-      let obsFloat = Tensor<Float>(obsDouble).reshaped(to: [1, 4])
-      let actionProbabilities = softmax(net.applied(to: obsFloat, in: context))
-
-      let actProbs = actionProbabilities[0].makeNumpyArray()
-      let len = Python.len(actProbs)
-      assert(actionCount == scalarFromPython(value: len) as Int32)
+      let observationPython = Tensor<Double>(numpy: observationNumpy).unwrapped()
+      let actionProbabilities = softmax(
+        net.applied(to: Tensor<Float>(observationPython).reshaped(to: [1, 4]),
+                    in: context))
```
Comment (rxwei) [inline Gym/CartPole.swift]: * Each line break indents by 4 spaces from the previous line.
* Unnecessary generic parameter.
* You can fit all of this in two lines because we are applying a 100 column limit. I recommend loosening line wrapping elsewhere in this file as well.
```swift
        let actionProbabilities =
            softmax(net.applied(to: Tensor(observationPython).reshaped(to: [1, 4]), in: context))
```

File: Gym/CartPole.swift [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+      let (nextObservation, reward, isDone, _) = env.step(actionPython).tuple4
+      // print(nextObservation)
+      // print(reward)
+
+      steps.append(Episode.Step(observation: Tensor<Float>(observationPython),
+                                action: Int32(actionPython).unwrapped()))
+
+      episodeReward += Float(reward).unwrapped()
+
+      if Bool(isDone).unwrapped() {
```
Comment (rxwei) [inline Gym/CartPole.swift]:
```suggestion
      if isDone == true {
```

File: Gym/CartPole.swift [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+
+  return episodes
+}
+
+let env = gym.make("CartPole-v0")
+let observationSize = Int32(env.observation_space.shape[0]).unwrapped()
+let actionCount = Int32(env.action_space.n).unwrapped()
+// print(actionCount)
+
+var net = Net(observationSize: Int(observationSize), hiddenSize: hiddenSize, actionCount: Int(actionCount))
```
Comment (rxwei) [inline Gym/CartPole.swift]: Why not make this `Int` in the first place? Same for other Int32->Int conversions.

File: Gym/CartPole.swift (hunk 16 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 17 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 18 above)
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift (hunk 19 above)
Comment (mhong) [reply Gym/CartPole.swift]: Done, and converted some other `Int32` to `Int` as I can.

**Comment by marcrasi** [comment]
Here's a copy of this that runs in Colab: https://colab.research.google.com/drive/1PzJGhCiHrgVWQJZDOlVRP_LpGiskII9-

I made zero modifications to the code, and it runs without problems. I did split it up into a few logical cells, though it also works to just paste the whole file into one cell.

**Comment by mhong** [comment]
Great, work, @marcrasi!

@rxwei, is this PR good to submit? Thank you both.

File: Gym/CartPole.swift [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+    // print("target tensor has shape \(target.shapeTensor)")
+
+    totalReward += episode.reward
+    retainedEpisodeCount += 1
+  }
+  
+  return (input, target, retainedEpisodeCount, totalReward / Float(retainedEpisodeCount))
+}
+
+func nextBatch(env: PythonObject,
```
[the remark this answers is not in the document]
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+  return episodes
+}
+
+let env = gym.make("CartPole-v0")
+let observationSize = Int(env.observation_space.shape[0]).unwrapped()
+let actionCount = Int(env.action_space.n).unwrapped()
+// print(actionCount)
+
+var net = Net(observationSize: Int(observationSize), hiddenSize: hiddenSize, actionCount: actionCount)
+/// SGD optimizer reaches convergence with ~125 mini batches, while Adam uses ~25.
```
[the remark this answers is not in the document]
Comment (mhong) [reply Gym/CartPole.swift]: done

File: Gym/CartPole.swift [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+
+  let steps: [Step]
+  let reward: Float
+}
+
+/// Filter out bad/short episodes before we feed them as neural net training
+/// data.
+func filteringBatch(episodes: [Episode],
+                    actionCount: Int
+) -> (input: Tensor<Float>,
```
[the remark this answers is not in the document]
Comment (mhong) [reply Gym/CartPole.swift]: Done

**Comment by rxwei** [comment]
Actually this was not switched to 4 space indentation. @mhong can you do that when you get a chance?

**Comment by mhong** [comment]
@rxwei: Sent out https://github.com/tensorflow/swift-models/pull/54 for 4 space indentation. Thanks.
